libraries/rib-base/src/test/java/com/badoo/ribs/routing/source/impl/PoolTest.kt

```diff
@@ -0,0 +1,175 @@
+    }
+
+    @Test
+    fun `GIVEN conf1 is active WHEN deactivate is called for conf1 THEN conf1 is inactive`() {
+        val pool = pool()
+
+        pool.deactivate(conf1)
+
+        assertEquals(false, routingHistoryElements.isActive(Configuration.Conf1))
+    }
+
+    @Test
+    fun `GIVEN conf1 is active WHEN deactivate is called for conf2 THEN conf1 is active`() {
+        val pool = pool(
+            listOf(
+                Item(
+                    configuration = Configuration.Conf1,
+                    identifier = conf1,
+                    isActive = true
+                ),
+                Item(
+                    configuration = Configuration.Conf2,
+                    identifier = conf2,
+                    isActive = true
+                )
+            )
+        )
+
+        pool.deactivate(conf2)
+
+        assertEquals(true, routingHistoryElements.isActive(Configuration.Conf1))
+    }
+
+    @Test
+    fun `GIVEN pool contains conf1 and conf2 WHEN remove is called for conf2 THEN pool contains only conf1`() {
+        val conf1Item = Item(
+            configuration = Configuration.Conf1,
+            identifier = conf1,
+            isActive = true
+        )
+        val conf2Item = Item(
+            configuration = Configuration.Conf2,
+            identifier = conf2,
+            isActive = true
+        )
+        val pool = pool(listOf(conf1Item, conf2Item))
+
+        pool.remove(conf2)
+
+        assertEquals(listOf(conf1Item.toRoutingHistoryElement()), routingHistoryElements)
+    }
+
+    @Test
+    fun `GIVEN pool contains conf1 WHEN add is called for conf2 THEN pool contains conf1 and conf2`() {
+        val conf1Item = Item(
+            configuration = Configuration.Conf1,
+            identifier = conf1,
+            isActive = true
+        )
+        val pool = pool(listOf(conf1Item))
+
+        val conf2Item = Item(
+            configuration = Configuration.Conf2,
+            identifier = conf2,
+            isActive = true
+        )
+        pool.add(conf2Item)
+
+        assertEquals(listOf(conf1Item.toRoutingHistoryElement(), conf2Item.toRoutingHistoryElement()), routingHistoryElements)
+    }
+
+    private fun pool(
+        initialItems: List<Item<Configuration>> =
+            listOf(
+                Item(
+                    configuration = Configuration.Conf1,
+                    identifier = conf1,
+                    isActive = true
+                ),
+                Item(
+                    configuration = Configuration.Conf2,
+                    identifier = conf2,
+                    isActive = false
+                )
+            )
+    ): Pool<Configuration> =
+        Pool(
+            initialItems = initialItems,
+            allowRepeatingConfigurations = false,
+            buildParams = emptyBuildParams()
+        ).apply {
+            observe { routingHistory -> routingHistoryElements = routingHistory.toList() }
```

Comment:
BTW you do not verify that you do not have multiple observer notifications, the thing that you fixed by introducing separate `updateState` method.